pkg/controller/sync/aggregator/aggregator_test.go

```diff
@@ -0,0 +1,237 @@
+							Group:   "group2",
+							Version: "v1",
+							Kind:    "Kind2",
+						},
+					},
+				},
+			},
+			expectAdded: true,
+			expectData: map[Key]map[schema.GroupVersionKind]struct{}{
+				{
+					Kind: syncset,
+					Name: "foo",
+				}: {
+					{
+						Group:   "group1",
+						Version: "v1",
+						Kind:    "Kind1",
+					}: {},
+					{
+						Group:   "group2",
+						Version: "v1",
+						Kind:    "Kind2",
+					}: {},
+				},
+			},
+			expectRev: map[schema.GroupVersionKind]map[Key]struct{}{
+				{
+					Group:   "group1",
+					Version: "v1",
+					Kind:    "Kind1",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+				},
+				{
+					Group:   "group2",
+					Version: "v1",
+					Kind:    "Kind2",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+				},
+			},
+		},
+		{
+			name: "add two keys and GVKs",
+			keyGVKs: []upsertKeyGVKs{
+				{
+					key: Key{
+						Kind: syncset,
+						Name: "foo",
+					},
+					gvks: []schema.GroupVersionKind{
+						{
+							Group:   "group1",
+							Version: "v1",
+							Kind:    "Kind1",
+						},
+						{
+							Group:   "group2",
+							Version: "v1",
+							Kind:    "Kind2",
+						},
+					},
+				},
+				{
+					key: Key{
+						Kind: configsync,
+						Name: "foo",
+					},
+					gvks: []schema.GroupVersionKind{
+						{
+							Group:   "group1",
+							Version: "v1",
+							Kind:    "Kind1",
+						},
+						{
+							Group:   "group2",
+							Version: "v1",
+							Kind:    "Kind2",
+						},
+					},
+				},
+			},
+			expectAdded: true,
+			expectData: map[Key]map[schema.GroupVersionKind]struct{}{
+				{
+					Kind: syncset,
+					Name: "foo",
+				}: {
+					{
+						Group:   "group1",
+						Version: "v1",
+						Kind:    "Kind1",
+					}: {},
+					{
+						Group:   "group2",
+						Version: "v1",
+						Kind:    "Kind2",
+					}: {},
+				},
+				{
+					Kind: configsync,
+					Name: "foo",
+				}: {
+					{
+						Group:   "group1",
+						Version: "v1",
+						Kind:    "Kind1",
+					}: {},
+					{
+						Group:   "group2",
+						Version: "v1",
+						Kind:    "Kind2",
+					}: {},
+				},
+			},
+			expectRev: map[schema.GroupVersionKind]map[Key]struct{}{
+				{
+					Group:   "group1",
+					Version: "v1",
+					Kind:    "Kind1",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+					{
+						Kind: configsync,
+						Name: "foo",
+					}: {},
+				},
+				{
+					Group:   "group2",
+					Version: "v1",
+					Kind:    "Kind2",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+					{
+						Kind: configsync,
+						Name: "foo",
+					}: {},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			agg := NewGVKAggregator()
+
+			for _, keyGVKs := range tt.keyGVKs {
+				added, err := agg.UpsertWithValidation(keyGVKs.key, keyGVKs.gvks)
+
+				if tt.expectErr {
+					assert.Error(t, err, "expected an error but none occurred")
```

Comment:
If we fail to add an object, the rest of the test is probably moot.  Should we do a `Fatal` instead of an `Error`?

Reply:
so `assert.Error` fails if we don't get an error. But if we do get an error, we pass the check and `return` because we know as test writers that the rest of the checks are not applicable